### PR TITLE
Global state handler

### PR DESCRIPTION
**Pull request by viacheslav-fomin-main**
Fixes: https://github.com/verygoodsecurity/vault/issues/1934

### CODE REVIEW

File: src/main/java/org/littleshoot/proxy/CustomGlobalState.java [hunk 1]
```diff
@@ -0,0 +1,18 @@
+package org.littleshoot.proxy;
+
+import io.netty.channel.Channel;
+
+public interface CustomGlobalState {
```
Comment (mjallday) [inline src/main/java/org/littleshoot/proxy/CustomGlobalState.java]: `GlobalStateHandler`

File: src/main/java/org/littleshoot/proxy/impl/ProxyConnection.java [hunk 2]
```diff
@@ -742,6 +742,43 @@ public void channelRead(ChannelHandlerContext ctx, Object msg)
...
+        @Override
+        public void channelRead(ChannelHandlerContext ctx, Object msg)
+            throws Exception {
+            proxyServer.getCustomGlobalState().restore(clientToProxyConnection.channel);
+            super.channelRead(ctx, msg);
+        }
+    }
+
+    @Sharable
+    protected class OutboundGlobalState extends
```
Comment (mjallday) [inline src/main/java/org/littleshoot/proxy/impl/ProxyConnection.java]: `OutboundGlobalStateAdapter`

File: pom.xml [hunk 3]
```diff
@@ -3,7 +3,7 @@
     <groupId>org.littleshoot</groupId>
     <artifactId>littleproxy</artifactId>
     <packaging>jar</packaging>
-    <version>1.1.3.3-VGS-SNAPSHOT</version>
+    <version>1.1.3.4-VGS-SNAPSHOT</version>
```
Comment (mjallday) [inline pom.xml]: this is new functionality. i think you should be bumping the minor version number

File: src/main/java/org/littleshoot/proxy/impl/ClientToProxyConnection.java [hunk 4]
```diff
@@ -770,6 +771,15 @@ protected void exceptionCaught(Throwable cause) {
...
 
+    @Override
+    public void channelRegistered(ChannelHandlerContext ctx) throws Exception {
+        try {
+            this.proxyServer.getGlobalStateHandler().persistToChannel(ctx.channel());
+        } finally {
+            super.channelRegistered(ctx);
+        }
+    }
+
```
Comment (viacheslav-fomin-main) [inline src/main/java/org/littleshoot/proxy/impl/ClientToProxyConnection.java:784]: This is a client to proxy connection. A moment where a client connects to the proxy and a channel is registered. We can generate a trace id here and put it to channel attributes so we can restore it into MDC later.

File: src/main/java/org/littleshoot/proxy/impl/ClientToProxyConnection.java [hunk 5]
```diff
@@ -817,7 +831,12 @@ private void initChannelPipeline(ChannelPipeline pipeline) {
                 new IdleStateHandler(0, 0, proxyServer
                         .getIdleConnectionTimeout()));
 
+        if (proxyServer.getGlobalStateHandler() != null) {
+            pipeline.addLast("outboundGlobalStateHandler", new OutboundGlobalStateHandler(this));
+        }
+
         pipeline.addLast("handler", this);
+
     }
```
Comment (viacheslav-fomin-main) [inline src/main/java/org/littleshoot/proxy/impl/ClientToProxyConnection.java:842]: `OutboundGlobalStateHandler` is placed after the `handler` as the `handler` will not take part in outbound message processing as it is of inbound type. It just writes data to the channel which starts the outbound processing where `OutboundGlobalStateHandler ` is the first in the chain.

File: src/main/java/org/littleshoot/proxy/impl/ClientToProxyConnection.java [hunk 6]
```diff
@@ -791,6 +801,10 @@ protected void exceptionCaught(Throwable cause) {
     private void initChannelPipeline(ChannelPipeline pipeline) {
         LOG.debug("Configuring ChannelPipeline");
 
+        if (proxyServer.getGlobalStateHandler() != null) {
+            pipeline.addLast("inboundGlobalStateHandler", new InboundGlobalStateHandler(this));
+        }
+
         pipeline.addLast("bytesReadMonitor", bytesReadMonitor);
```
Comment (viacheslav-fomin-main) [inline src/main/java/org/littleshoot/proxy/impl/ClientToProxyConnection.java:810]: `InboundGlobalStateHandler ` placed at the very beginning

File: src/main/java/org/littleshoot/proxy/impl/ProxyConnection.java [hunk 7]
```diff
@@ -742,6 +743,46 @@ public void channelRead(ChannelHandlerContext ctx, Object msg)
...
+        }
+
+        @Override
+        public void channelRead(ChannelHandlerContext ctx, Object msg)
+            throws Exception {
+            proxyServer.getGlobalStateHandler().restoreFromChannel(clientToProxyConnection.channel);
+            super.channelRead(ctx, msg);
+            proxyServer.getGlobalStateHandler().clear();
+        }
+    }
```
Comment (viacheslav-fomin-main) [inline src/main/java/org/littleshoot/proxy/impl/ProxyConnection.java]: `super.channelRead(ctx, msg);` takes the complete pipeline. Inside it does data processing and then calls `fireChannelRead` which takes a next handler in the chain and calls the same `channelRead` on it after which the process repeats until the last handler is reached. After that the method `super.channelRead(ctx, msg);` is returned and re clear the context.

File: src/main/java/org/littleshoot/proxy/HttpProxyServerBootstrap.java [hunk 8]
```diff
@@ -213,6 +213,21 @@ HttpProxyServerBootstrap withProxyToServerExHandler(
 
     /**
      * <p>
+     * Specify an {@link GlobalStateHandler} to customize a global state based on channel attributes
+     * </p>
+     *
+     * <p>
+     * Default = null
```
Comment (osklyarenko) [inline src/main/java/org/littleshoot/proxy/HttpProxyServerBootstrap.java:220]: let's not use null. pls make a default impl that is does nothing.

File: pom.xml [hunk 9]
```diff
@@ -3,7 +3,7 @@
     <groupId>org.littleshoot</groupId>
     <artifactId>littleproxy</artifactId>
     <packaging>jar</packaging>
-    <version>1.1.3.3-VGS-SNAPSHOT</version>
+    <version>1.1.4.3-VGS-SNAPSHOT</version>
```
Comment (osklyarenko) [inline pom.xml]: should be `1.1.4.0`

File: src/main/java/org/littleshoot/proxy/impl/ProxyConnection.java [hunk 10]
```diff
@@ -742,6 +743,46 @@ public void channelRead(ChannelHandlerContext ctx, Object msg)
...
+        InboundGlobalStateHandler(ProxyConnection<HttpRequest> clientToProxyConnection) {
+            this.clientToProxyConnection = clientToProxyConnection;
+        }
+
+        @Override
+        public void channelRead(ChannelHandlerContext ctx, Object msg)
+            throws Exception {
+            proxyServer.getGlobalStateHandler().restoreFromChannel(clientToProxyConnection.channel);
+            super.channelRead(ctx, msg);
+            proxyServer.getGlobalStateHandler().clear();
```
Comment (osklyarenko) [inline src/main/java/org/littleshoot/proxy/impl/ProxyConnection.java]: pls move it to a `finally` block

**Review by osklyarenko — APPROVED** [review]

**Review by k-sever — APPROVED** [review]
LGTM

**Review by mjallday — COMMENTED** [review]
recommend you add your comments on the placement of the handlers within the pipeline to the source code so it's obvious what it does.

should we also make a pr against the main lp repo?

**Comment by viacheslav-fomin-main** [comment]
@mjallday sure, will add the comments after we test all this with proper changes in vault. 
Not sure if this change is needed in the main repo. This still looks more like a hack..

**Pull request merged**